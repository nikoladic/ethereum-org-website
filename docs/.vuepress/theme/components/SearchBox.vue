<template>
  <!-- 
    Our parent div watches to see if any children are focused
    unFocus checks if the next target is a child before unfocusing
   -->
  <div
    :class="searchClasses"
    @focusin="focused = true"
    @focusout="unFocus"
    @keydown.esc="unFocus"
  >
    <h1 class="search-title l3 mt-0 flex flex-center a-breakm-hidden">
      <icon
        name="chevron-right"
        class="icon-back"
        @click.native="$emit('search-toggle')"
        @keyup.enter="$emit('search-toggle')"
      />
      Search
    </h1>

    <div class="search-bar">
      <input
        id="main-search-field"
        @input="query = $event.target.value"
        aria-label="Search"
        :value="query"
        autocomplete="off"
        spellcheck="false"
        @keyup.enter="go(focusIndex)"
        placeholder="Search"
      />
      <icon name="search" class="icon-search-field" />
    </div>

    <div v-if="blankState" class="blank-state">
      <div class="blank-state-emoji">{{ blankState.emoji }}</div>
      <span>{{ blankState.text }}</span>
    </div>
    <template v-else>
      <h2 v-if="!blankState" class="results-title l4 a-breakm-hidden">
        Results
      </h2>
      <ul
        v-if="!blankState"
        class="suggestions pl-0 mt-0 no-list a-breakm-absolute"
      >
        <li v-for="(s, i) in suggestions">
          <router-link
            :to="s.path"
            class="result-link pa-05 flex flex-column align-center"
            @mousedown.native="
              $router.push(s.path),
                $emit('search-toggle'),
                $emit('nav-toggle', false),
                forceUnFocus()
            "
            @keyup.enter="
              $emit('search-toggle'), $emit('nav-toggle', false), forceUnFocus()
            "
          >
            <span v-if="s.header" class="result-title mb-025 tc-text400">{{
              s.header.title
            }}</span>
            <span class="result-page tc-text100">{{ s.title || s.path }}</span>
          </router-link>
        </li>
      </ul>
    </template>
  </div>
</template>

<script>
import { resolveHeaderTitle } from '../utils/util'

export default {
  props: {
    isSearchVisible: {
      type: Boolean,
      default: false
    },
    method: { type: Function }
  },

  data() {
    return {
      query: '',
      focusIndex: 0,
      focused: false
    }
  },

  watch: {
    // clear query when search hidden
    isSearchVisible: function() {
      !this.isSearchVisible && (this.query = '')
    }
  },
  computed: {
    searchClasses() {
      return {
        'search-box': true,
        'hidden a-breakm-block': !this.isSearchVisible,
        'focus-within': this.focused
      }
    },

    blankState() {
      if (!this.query.length) {
        return { emoji: '⛵️', text: 'Search away!' }
      } else if (!this.suggestions.length) {
        return { emoji: '😕', text: 'No results found' }
      } else {
        return false
      }
    },

    suggestions() {
      const query = this.query.trim().toLowerCase()
      if (!query) {
        return
      }

      const { pages, themeConfig } = this.$site
      const max = themeConfig.searchMaxSuggestions || 5
      const localePath = this.$localePath
      const matches = item =>
        item.title && item.title.toLowerCase().indexOf(query) > -1
      const res = []
      for (let i = 0; i < pages.length; i++) {
        if (res.length >= max) break
        const p = pages[i]
        // filter out results that do not match current locale
        if (this.getPageLocalePath(p) !== localePath) {
          continue
        }
        if (matches(p)) {
          res.push(p)
        } else if (p.headers) {
          for (let j = 0; j < p.headers.length; j++) {
            if (res.length >= max) break
            const h = p.headers[j]
            if (matches(h)) {
              h.title = resolveHeaderTitle(h.title)
              res.push(
                Object.assign({}, p, {
                  path: p.path + '#' + h.slug,
                  header: h
                })
              )
            }
          }
        }
      }
      return res
    }
  },

  methods: {
    unFocus(e) {
      e.relatedTarget
        ? !e.relatedTarget.classList.contains('result-link') &&
          (this.focused = false)
        : (this.focused = false)
      e.target.blur()
    },
    forceUnFocus() {
      this.focused = false
    },
    getPageLocalePath(page) {
      for (const localePath in this.$site.locales || {}) {
        if (localePath !== '/' && page.path.indexOf(localePath) === 0) {
          return localePath
        }
      }
      return '/'
    },

    go(i) {
      if (!this.suggestions) {
        return
      }
      this.$router.push(this.suggestions[i].path)
      this.query = ''
      this.focusIndex = 0
    }
  }
}
</script>

<style lang="stylus">
@import '../styles/config.styl'

.search-title
  line-height 1

.search-bar
  position relative

.search-box
  position absolute
  top unquote('calc( -100 * var(--vh) + ' + $mobileBottomDrawerHeight + ')')
  left 0
  right 0
  height unquote('calc(100 * var(--vh))')
  box-sizing: border box
  padding 1em
  transition all 0.25s ease-in-out
  display flex
  flex-direction column

  &, *, *:before, *:after
    box-sizing border-box

  input
    appearance none
    border none
    outline none
    font-size 1rem
    height auto
    line-height 1.4
    padding .5rem 2rem .5rem .5rem
    border-radius 0.25em
    width 100%

    &:focus
      cursor auto
      appearance none
      outline none

  *, *:before, *:after
    box-sizing: border box

.icon-search-field
  position absolute
  top 50%
  margin-top -12px
  right 6px

.icon-back
  transform rotate(180deg)
  margin-right .5em

.blank-state
  display flex
  flex-direction column
  align-items center
  justify-content center
  margin-top 10vw
  align-self center
  width unquote('min(60vw, 280px)')
  height unquote('min(60vw, 280px)')
  border-radius 100%

.blank-state-emoji
  height  80px
  line-height 1
  font-size: 80px

.search-hidden
  display none

.result-link
  min-height 2em
  border-radius 0.4em
  margin 0 -0.5em

.result-title
  margin-bottom 0.25em

@media (min-width: $breakM)
  .search-title,
  .results-title,
  .icon-back,
  .blank-state,
  .suggestions
    display none

  .icon-back
    display none
  .search-box
    display inline-block
    width: auto
    position relative
    background transparent
    padding 0
    top 0
    height initial

    &.focus-within
      .suggestions
        display block
      .blank-state, .suggestions
        display flex

  .search-hidden
      transform none

  .suggestions, .blank-state
    margin 0
    flex-direction column
    width 120%
    position absolute
    top calc(100% + 4px)
    border-radius 0.25em
    border-radius 0.25em
  .result-link
    padding 0.5em
    margin 0
    border-radius 0

// Light Mode
.search-box
  background $colorWhite500
  input
    border 1px solid $colorBlack50
.result-link
  &:hover, &:focus
    background: alpha($colorPrimary100, 0.2)

.result-title, .result-page
  color: $colorBlack500
.result-title + .result-page
  color: $colorBlack100

.blank-state
  color $colorBlack50
  background $colorWhite600

@media (min-width: $breakM)
  .suggestions, .blank-state
    background $colorWhite500
    border 1px solid $colorWhite800

// Dark Mode
.dark-mode
  .search-box
    background $colorBlack500
    input
      color $colorWhite600
      background $colorBlack200
      border 1px solid $colorWhite900
  .result-title, .result-page
    color: $colorWhite500
  .result-title + .result-page
    color: $colorWhite900
  .blank-state
    background $colorBlack300
  @media (min-width: $breakM)
    .suggestions, .blank-state
      background $colorBlack300
      border 1px solid $colorBlack100
</style>
