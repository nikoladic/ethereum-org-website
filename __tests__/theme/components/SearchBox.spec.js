import { shallowMount } from '@vue/test-utils'
import SearchBox from "@/theme/components/SearchBox.vue"


describe("As user i see a searchbox", () => {
  test('empty, without content', () => {
    const wrapper = shallowMount(SearchBox, {
      mocks: {
        $lang: 'en-US',
        $localePath: '/',
        $page: {path: '/'},
        $pages: ['/'],
        $site: {
          themeConfig: {
            search: true,
            searchMaxSuggestions: 10
          },
          locales: {
            "es": {label: 'Español', lang: 'es'},
            "en": {label: 'English', lang: 'en-US'}
          },
          nav: []
        },
        $router: {
          options: {routes: []}
        }
      }
    });

    expect(wrapper.element).toMatchSnapshot()
  });


  test('with text and recomendations ', () => {
    const wrapper = shallowMount(SearchBox, {
      mocks: {
        $lang: 'en-US',
        $localePath: '/',
        $page: {path: '/'},
        $site: {
          pages: [{title: 'Whales everywhere!', headers: ['Whale', 'Crypto', 'Contracts']}],
          themeConfig: {
            search: true,
            searchMaxSuggestions: 10
          },
          locales: {
            "/es/": {label: 'Español', lang: 'es'},
            "/": {label: 'English', lang: 'en-US'}
          },
          nav: []
        },
        $router: {
          options: {routes: []}
        }
      }
    });

    wrapper.find("input").setValue("Whale")
    expect(wrapper.element).toMatchSnapshot()
  });


  test('with text but without recommendations', () => {
    const wrapper = shallowMount(SearchBox, {
      mocks: {
        $lang: 'en-US',
        $localePath: '/',
        $page: {path: '/'},
        $site: {
          pages: [{title: 'Whales everywhere!', headers: ['Whale', 'Crypto', 'Contracts']}],
          themeConfig: {
            search: true,
            searchMaxSuggestions: 10
          },
          locales: {
            "/es/": {label: 'Español', lang: 'es'},
            "/": {label: 'English', lang: 'en-US'}
          },
          nav: []
        },
        $router: {
          options: {routes: []}
        }
      }
    });

    wrapper.find("input").setValue("Smart")
    expect(wrapper.element).toMatchSnapshot()
  });
});