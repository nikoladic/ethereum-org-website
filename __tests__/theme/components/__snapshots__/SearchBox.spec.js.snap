// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`As user i see a searchbox empty, without content 1`] = `
<div
  class="search-box"
>
  <input
    aria-label="Search"
    autocomplete="off"
    class=""
    spellcheck="false"
  />
   
  <!---->
</div>
`;

exports[`As user i see a searchbox with text and recomendations  1`] = `
<div
  class="search-box"
>
  <input
    aria-label="Search"
    autocomplete="off"
    class=""
    spellcheck="false"
  />
   
  <!---->
</div>
`;

exports[`As user i see a searchbox with text but without recommendations 1`] = `
<div
  class="search-box"
>
  <input
    aria-label="Search"
    autocomplete="off"
    class=""
    spellcheck="false"
  />
   
  <!---->
</div>
`;
